src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFile.cs

```diff
@@ -166,16 +167,11 @@ public bool Equals(LockFile other)
 
             return IsLocked == other.IsLocked
                 && Version == other.Version
-                && ProjectFileDependencyGroups.OrderBy(group => group.FrameworkName, StringComparer.OrdinalIgnoreCase)
-                    .SequenceEqual(other.ProjectFileDependencyGroups.OrderBy(
-                        group => group.FrameworkName, StringComparer.OrdinalIgnoreCase))
-                && Libraries.OrderBy(library => library.Name, StringComparer.OrdinalIgnoreCase)
-                    .SequenceEqual(other.Libraries.OrderBy(library => library.Name, StringComparer.OrdinalIgnoreCase))
-                && Targets.OrderBy(target => target.Name).SequenceEqual(other.Targets.OrderBy(target => target.Name))
-                && ProjectFileToolGroups.OrderBy(group => group.FrameworkName, StringComparer.OrdinalIgnoreCase)
-                    .SequenceEqual(other.ProjectFileToolGroups.OrderBy(
-                        group => group.FrameworkName, StringComparer.OrdinalIgnoreCase))
-                && Tools.OrderBy(target => target.Name).SequenceEqual(other.Tools.OrderBy(target => target.Name));
+                && ProjectFileDependencyGroups.OrderedEquals(other.ProjectFileDependencyGroups, group => group.FrameworkName, StringComparer.OrdinalIgnoreCase)
+                && Libraries.OrderedEquals(other.Libraries, library => library.Name, StringComparer.OrdinalIgnoreCase)
+                && Targets.OrderedEquals(other.Targets, target => target.Name)
```

Comment:
There is at least one scenario where the field we are ordering on is not a string. For example, [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.RuntimeModel/CompatibilityProfile.cs#L43), where we are sorting by `FrameworkRuntimePair`, which is an `IComparable`.

One alternative would be to throw if `TSource` or `TKey` is of type string and the corresponding comparer is not provided, though that doesn't provide compile-time verification.